example/memory.cc

```diff
@@ -72,80 +54,61 @@ void run() {
-  std::cout << "Compiling module..." << std::endl;
-  auto module = wasm::Module::make(store, binary);
-  if (!module) {
-    std::cout << "> Error compiling module!" << std::endl;
-    exit(1);
-  }
+  auto file = load_file("memory.wasm");
 
   // Instantiate.
   std::cout << "Instantiating module..." << std::endl;
-  auto imports = wasm::vec<wasm::Extern*>::make();
-  auto instance = wasm::Instance::make(store, module.get(), imports);
-  if (!instance) {
-    std::cout << "> Error instantiating module!" << std::endl;
+  auto instance = store.instantiate(file, wasm::Imports{});
+  if (auto error = instance.error()) {
+    std::cout << "> Error instantiating module: " << error->what() << std::endl;
     exit(1);
   }
 
   // Extract export.
   std::cout << "Extracting exports..." << std::endl;
-  auto exports = instance->exports();
   size_t i = 0;
-  auto memory = get_export_memory(exports, i++);
-  auto size_func = get_export_func(exports, i++);
-  auto load_func = get_export_func(exports, i++);
-  auto store_func = get_export_func(exports, i++);
+  auto memory = instance->exported_memory(0);
+  auto size_func = instance->exported_func<int32_t()>(i++);
+  auto load_func = instance->exported_func<int32_t(int32_t)>(i++);
+  auto store_func = instance->exported_func<int32_t(int32_t)>(i++);
 
   // Try cloning.
-  assert(memory->copy()->same(memory));
+  // ???
+  //assert(memory->copy()->same(memory));
 
   // Check initial memory.
   std::cout << "Checking memory..." << std::endl;
-  check(memory->size(), 2u);
-  check(memory->data_size(), 0x20000u);
-  check(memory->data()[0], 0);
-  check(memory->data()[0x1000], 1);
-  check(memory->data()[0x1003], 4);
-
-  check(call(size_func), 2);
-  check(call(load_func, 0), 0);
-  check(call(load_func, 0x1000), 1);
-  check(call(load_func, 0x1003), 4);
-  check(call(load_func, 0x1ffff), 0);
+  check(memory->size(), wasm::Pages(2));
+  check(memory->contents()->size(), 0x20000u);
+  check(memory->contents()[0], 0);
+  check(memory->contents()[0x1000], 1);
+  check(memory->contents()[0x1003], 4);
+
+  check(size_func(), 2);
+  check(load_func(0), 0);
+  check(load_func(0x1000), 1);
+  check(load_func(0x1003), 4);
+  check(load_func(0x1ffff), 0);
   check_trap(load_func, 0x20000);
 
   // Mutate memory.
   std::cout << "Mutating memory..." << std::endl;
-  memory->data()[0x1003] = 5;
+  memory->contents()[0x1003] = 5;
   check_ok(store_func, 0x1002, 6);
   check_trap(store_func, 0x20000, 0);
 
-  check(memory->data()[0x1002], 6);
-  check(memory->data()[0x1003], 5);
-  check(call(load_func, 0x1002), 6);
-  check(call(load_func, 0x1003), 5);
+  check(memory->contents()[0x1002], 6);
+  check(memory->contents()[0x1003], 5);
+  check(load_func(0x1002), 6);
+  check(load_func(0x1003), 5);
 
   // Grow memory.
   std::cout << "Growing memory..." << std::endl;
   check(memory->grow(1), true);
   check(memory->size(), 3u);
-  check(memory->data_size(), 0x30000u);
+  check(memory->contents()->size(), 0x30000u);
```

Comment:
Hmm, what's the idea here? What is `contents`, is it type that can be used on its own?

Reply:
Ideally it'd be a `std::span<char>` but I realized that std::span didn't make it into c++17.

Thinking about this over the weekend, I now think we should provide a span-compatible type for C++17 users but under a macro that uses just makes our wasm::span = std::span when the user has std::span.

Reference: https://en.cppreference.com/w/cpp/container/span